### DISCUSSION TITLE
py-capellambse and state machines

### DISCUSSION THREAD

**Discussion by khammoutene** [dbinfrago/py-capellambse#170] [Q&A] [answered]
I am modelling states machines and I am trying to access to the “effects” using py-capellambse.

I have modelled both an effect in a transition and in a state.

Capella is restricting me to Service elements to describe effects that’s why you see “Service 3” in the diagram. I was expecting to be able refer to functions for an effect but this is not my question to you
(I have created a topic discussion in: [https://forum.mbse-capella.org/t/modelling-effects-in-state-transitions-in-a-state-machine/5833]
 
This is a basic model:

![image](https://user-images.githubusercontent.com/27721103/198000730-d6c03a75-ceb3-42fc-8649-50bece2000ad.png)

My question to you is that I cannot see “Service 3”, i.e. effects, in both the state information and transition information that I am accessing via py-capellambse.
 
 
<State 'Draft' (dd843949-ae9b-4161-bd2d-b53c239feb07)>
.constraints = []
.description = None
.diagrams = []
.functions = []
.name = 'Draft'
.parent = <Region 'Default Region' (c417fb92-670a-4f03-b782-cb7504368856)>
.progress_status = 'NOT_SET'
.realized_states = [0] <State 'Draft' (c127914c-0b8b-42ee-b3b6-510482a2df74)>
.realizing_states = []
.regions = [0] <Region 'region' (30ecb727-0033-4582-8f36-4d639b79901c)>
.requirements = []
.summary = None
.traces = []
.uuid = 'dd843949-ae9b-4161-bd2d-b53c239feb07'
.xtype = 'org.polarsys.capella.core.data.capellacommon:State'
 
<StateTransition '' (571bd05e-8303-4fc2-89c9-4dac757a72e1)>
.constraints = []
.description = None
.destination = <State 'Draft' (dd843949-ae9b-4161-bd2d-b53c239feb07)>
.diagrams = []
.guard = None
.name = ''
.parent = <Region 'Default Region' (c417fb92-670a-4f03-b782-cb7504368856)>
.progress_status = 'NOT_SET'
.requirements = []
.source = <JoinPseudoState '' (5f210683-853e-4c76-a26c-a04c35ad4bbb)>
.summary = None
.traces = []
.triggers = [0] <Model element (org.polarsys.capella.core.data.capellacommon:ChangeEvent) 'Back to draft' (d7bab880-0620-4d88-b587-2306ddd0009e)>
.uuid = '571bd05e-8303-4fc2-89c9-4dac757a72e1'
.xtype = 'org.polarsys.capella.core.data.capellacommon:StateTransition'
 
Do you have any idea why?
 
Thank you,

**Comment by Wuestengecko** [comment, marked as answer]
Looks like we simply don't have what you're looking for yet. I've made a PR to add the `effects` attribute to StateTransition objects in #174.

**Reply by vik378** [reply]
I dont think it is correct to call this "effects": 
<img width="356" alt="image" src="https://user-images.githubusercontent.com/8280844/198272802-0b434f6f-0183-4cfc-a7c6-78d2ef9a350b.png">

These are activities that get executed on state entry / exit or while in state, so it should rather be
- `entry`
- `do_activity`
- `exit`

**Comment by vik378** [comment]
For StateTransition, `.effects` seem to be independent from the target state entry `entry` - effects is basically a thing on its own
![image](https://user-images.githubusercontent.com/8280844/198274406-bc8ebbb3-c064-46d8-b391-4be6cc73b254.png)

**Reply by khammoutene** [reply]
Thank you.

I don't want to start a methodological debate about actions, activities, effects, class operations, ....

I just need to access via py-capellambse anything I have entered or linked in the model.

What are the workarounds to access the "effects" entered in a state transition?

**Reply by ewuerger** [reply]
With the [latest release (v0.5.3)](https://github.com/DSD-DBS/py-capellambse/releases/tag/v0.5.3) you can just access `.effects` on any `StateTransition`. Kudos to @Wuestengecko.

**Reply by khammoutene** [reply]
Thank you and Kudos to @khammoutene to have failed unit tests.

**Reply by khammoutene** [reply]
Just tested the new version with an effect on a transition and it works. Tx.

It doesn't seem to work with states,  i.e. Entry; Do; Exit; Operational activities/Functions. Correct?

**Reply by ewuerger** [reply]
After reading your comment I implemented it with #181. (You can have a look into the [commit](https://github.com/DSD-DBS/py-capellambse/pull/181/commits/df3e67da01a380846243ca485bc1ecf0f2e162a2) to see how it was done. For now only the [API reference (generated documentation)](https://dsd-dbs.github.io/py-capellambse/code/capellambse.model.common.html#module-capellambse.model.common.accessors) is there for guidance.)

We'll add a page for attribute declaration via our concrete Accessors when you encounter missing ones. Ideally we'd like to generate the code against the meta model provided by the [official GitHub repo of capella](https://github.com/eclipse/capella) to reach full metamodel-coverage. Unfortunately advances in that direction weren't fruitful yet. Then you and others don't rely too much on the maintainers. But this will come in the next months b/c we are busy with internal projects right now. 

Edit: Add hint to concrete Accessors.

**Reply by khammoutene** [reply]
Thank you @ewuerger 

I will come back in few months.

**Reply by vik378** [reply]
well, I don't think it is months though.. our approach to coverage so far was (and will remain) reactive - when we or a external contributor find a missing thing we or external contributor propose a fix for it. We made a few attempts on 1:1 metamodel to code but we weren't convinced by the outcome. It is nice that you found missing bits on state machines, @khammoutene - this also goes along with one of the things from the methods backlog (stm to text) - and the fix for those 3 missing properties should be pretty trivial - maybe I draft it this evening and maybe it gets through the review sometimes this week or so.

**Reply by vik378** [reply]
oh, and I see that @ewuerger already did with https://github.com/DSD-DBS/py-capellambse/pull/181 🥳

**Reply by ewuerger** [reply]
> well, I don't think it is months though.. [...]

I was referring to the introductory page of  our concrete Accessors  when I wrote in the next months. I know atleast that it won't happen this november.

**Reply by khammoutene** [reply]
Thank you for your reactivity.

Do you have experience in developing Python code combining your approach with Py4C as I see value in both?
